integration/apidefinition_test.go

```diff
@@ -770,45 +771,230 @@ func TestApiDefinitionUpstreamCertificates(t *testing.T) {
 				calculatedCertID := string(tykOrg) + certFingerPrint
 
 				err = wait.For(func() (done bool, err error) {
-					hc := &http.Client{}
+					env := environmet.Env{}
+					env.Mode = v1alpha1.OperatorContextMode(mode)
+					env.Org = string(tykOrg)
+					env.Auth = string(tykAuth)
+					env.URL = tykConnectionURL
+
+					pkgContext := pkgclient.Context{
+						Env: env,
+						Log: log.NullLogger{},
+					}
 
-					req, err := http.NewRequest(
-						http.MethodGet,
-						fmt.Sprintf("%s/api/certs/?certId=%s&org_id=%s", dashboardLocalHost, calculatedCertID, string(tykOrg)),
-						nil,
-					)
+					// validate certificate was created
+					reqContext := pkgclient.SetContext(context.Background(), pkgContext)
+					exists := klient.Universal.Certificate().Exists(reqContext, calculatedCertID)
+
+					is.True(exists)
+
+					return true, nil
+				}, wait.WithTimeout(defaultWaitTimeout), wait.WithInterval(defaultWaitInterval))
+				is.NoErr(err)
+
+				return ctx
+			}).Feature()
+
+	testenv.Test(t, adCreate)
+}
+
+func TestApiDefinitionClientMTLS(t *testing.T) {
+	var (
+		apiDefClientMTLSWithCert    = "apidef-client-mtls-with-cert"
+		apiDefClientMTLSWithoutCert = "apidef-client-mtls-without-cert"
+		defaultVersion              = "Default"
+		opNs                        = "tyk-operator-system"
+		certName                    = "test-tls-secret-name"
+		tykConnectionURL            = ""
+	)
+
+	mode := os.Getenv("TYK_MODE")
+
+	switch mode {
+	case "pro":
+		tykConnectionURL = adminLocalhost
+	case "ce":
+		tykConnectionURL = gatewayLocalhost
+	}
+
+	testWithCert := features.New("Client MTLS with certificate").
+		Setup(func(ctx context.Context, t *testing.T, envConf *envconf.Config) context.Context {
+			testNS := ctx.Value(ctxNSKey).(string) //nolint: errcheck
+			is := is.New(t)
+			t.Log(testNS)
+
+			_, err := createTestTlsSecret(ctx, testNS, nil, envConf)
+			is.NoErr(err)
+
+			return ctx
+		}).
+		Setup(func(ctx context.Context, t *testing.T, envConf *envconf.Config) context.Context {
+			testNS := ctx.Value(ctxNSKey).(string) //nolint:errcheck
+			is := is.New(t)
+
+			// Create ApiDefinition with Client certificate
+			_, err := createTestAPIDef(ctx, testNS, func(apiDef *v1alpha1.ApiDefinition) {
+				apiDef.Name = apiDefClientMTLSWithCert
+				apiDef.Spec.UseMutualTLSAuth = true
+				apiDef.Spec.ClientCertificateRefs = []string{certName}
+				apiDef.Spec.VersionData.DefaultVersion = defaultVersion
+				apiDef.Spec.VersionData.NotVersioned = true
+				apiDef.Spec.VersionData.Versions = map[string]model.VersionInfo{
+					defaultVersion: {Name: defaultVersion},
+				}
+			}, envConf)
+			is.NoErr(err) // failed to create apiDefinition
+
+			return ctx
+		}).
+		Assess("API must have client certificate field defined",
```

Comment:
As far as I understand this test suite, it checks not only the existence of the certificate field, but it also validates the defined certificate. It might be better to rename this test or to separate it into two test cases, such as "check a field exists", "validation of secret" or something like this.

Reply:
It checks if certificate is uploaded and if client Certificate field was set.

Reply:
Makes sense. I will split it in two different Assess function.